File: sdk-tests/consul-go-tests/session_test.go
package tests

import (
	"testing"
	"time"

	"github.com/hashicorp/consul/api"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

// ==================== P0: Critical Tests ====================

// CS-001: Test session create with default validation
// Consul defaults: LockDelay=15s, Behavior="release", NodeChecks=["serfHealth"], TTL=""
func TestSessionCreateDefaults(t *testing.T) {
	client := getClient(t)

	// Create session with ONLY a name — all other fields should be defaulted
	sessionID, wm, err := client.Session().Create(&api.SessionEntry{
		Name: "test-defaults-" + randomID(),
	}, nil)
	require.NoError(t, err, "Session create should succeed")
	require.NotEmpty(t, sessionID, "Session ID must not be empty")
	assert.Greater(t, wm.RequestTime, time.Duration(0))
	defer client.Session().Destroy(sessionID, nil)

	// Get info and verify ALL defaults
	info, meta, err := client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	require.NotNil(t, info, "Session info must not be nil")

	// ID
	assert.Equal(t, sessionID, info.ID, "Session ID must match")

	// LockDelay default = 15s
	assert.Equal(t, 15*time.Second, info.LockDelay,
		"Default LockDelay must be 15s (Consul default)")

	// Behavior default = "release"
	assert.Equal(t, api.SessionBehaviorRelease, info.Behavior,
		"Default Behavior must be 'release' (Consul default)")

	// TTL default = "" (persistent session, no auto-expiry)
	assert.Empty(t, info.TTL,
		"Default TTL must be empty (persistent session, Consul default)")

	// NodeChecks default = ["serfHealth"]
	assert.Equal(t, []string{"serfHealth"}, info.NodeChecks,
		"Default NodeChecks must be ['serfHealth'] (Consul default)")

	// Node should be set
	assert.NotEmpty(t, info.Node, "Node must be set")

	// CreateIndex must be positive
	assert.Greater(t, info.CreateIndex, uint64(0))

	// QueryMeta
	assert.Greater(t, meta.LastIndex, uint64(0))
	assert.True(t, meta.KnownLeader)
}

// CS-002: Test session create with explicit TTL
func TestSessionCreateWithTTL(t *testing.T) {
	client := getClient(t)

	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name: "ttl-session-" + randomID(),
		TTL:  "30s",
	}, nil)
	require.NoError(t, err)
	defer client.Session().Destroy(sessionID, nil)

	info, _, err := client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	require.NotNil(t, info)

	// TTL should be set (Consul normalizes to "30s")
	assert.NotEmpty(t, info.TTL, "TTL must be set when explicitly provided")
	assert.Contains(t, info.TTL, "30", "TTL must contain '30'")
}

// CS-003: Test destroy session
func TestSessionDestroy(t *testing.T) {
	client := getClient(t)

	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name: "destroy-" + randomID(),
		TTL:  "30s",
	}, nil)
	require.NoError(t, err)

	// Destroy
	_, err = client.Session().Destroy(sessionID, nil)
	require.NoError(t, err, "Session destroy should succeed")

	// Verify gone
	info, _, err := client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	assert.Nil(t, info, "Destroyed session must return nil from Info()")
}

// CS-004: Test renew session
func TestSessionRenew(t *testing.T) {
	client := getClient(t)

	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name: "renew-" + randomID(),
		TTL:  "15s",
	}, nil)
	require.NoError(t, err)
	defer client.Session().Destroy(sessionID, nil)

	// Renew
	entry, _, err := client.Session().Renew(sessionID, nil)
	require.NoError(t, err, "Session renew should succeed")
	require.NotNil(t, entry, "Renewed session must not be nil")
	assert.Equal(t, sessionID, entry.ID, "Renewed session ID must match")
}

// CS-005: Test renew destroyed session returns nil
func TestSessionRenewDestroyed(t *testing.T) {
	client := getClient(t)

	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name: "renew-destroyed-" + randomID(),
		TTL:  "30s",
	}, nil)
	require.NoError(t, err)

	_, err = client.Session().Destroy(sessionID, nil)
	require.NoError(t, err)

	// Renew of destroyed session should return nil entry (not error)
	entry, _, err := client.Session().Renew(sessionID, nil)
	// Consul returns nil entry, some implementations may return error
	if err == nil {
		assert.Nil(t, entry,
			"Renew of destroyed session must return nil entry")
	}
}

// ==================== P1: Important Tests ====================

// CS-006: Test list sessions
func TestSessionList(t *testing.T) {
	client := getClient(t)

	// Create 3 sessions
	var ids []string
	for i := 0; i < 3; i++ {
		id, _, err := client.Session().Create(&api.SessionEntry{
			Name: "list-" + randomID(),
			TTL:  "30s",
		}, nil)
		require.NoError(t, err)
		ids = append(ids, id)
	}
	defer func() {
		for _, id := range ids {
			client.Session().Destroy(id, nil)
		}
	}()

	// List
	sessions, meta, err := client.Session().List(nil)
	require.NoError(t, err, "Session list should succeed")
	assert.GreaterOrEqual(t, len(sessions), 3, "Should have at least 3 sessions")
	assert.Greater(t, meta.LastIndex, uint64(0))

	// Verify our sessions are in the list
	foundCount := 0
	for _, s := range sessions {
		for _, id := range ids {
			if s.ID == id {
				foundCount++
				// Each session should have valid fields
				assert.NotEmpty(t, s.Node, "Session node must be set")
				assert.Greater(t, s.CreateIndex, uint64(0))
				break
			}
		}
	}
	assert.Equal(t, 3, foundCount, "All 3 created sessions must be found")
}

// CS-007: Test session with custom lock delay
func TestSessionLockDelay(t *testing.T) {
	client := getClient(t)

	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name:      "lockdelay-" + randomID(),
		TTL:       "30s",
		LockDelay: 5 * time.Second,
	}, nil)
	require.NoError(t, err)
	defer client.Session().Destroy(sessionID, nil)

	info, _, err := client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	require.NotNil(t, info)

	assert.Equal(t, 5*time.Second, info.LockDelay,
		"Custom LockDelay must be preserved")
}

// CS-008: Test session with delete behavior
func TestSessionDeleteBehavior(t *testing.T) {
	client := getClient(t)

	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name:     "delete-behavior-" + randomID(),
		TTL:      "30s",
		Behavior: api.SessionBehaviorDelete,
	}, nil)
	require.NoError(t, err)
	defer client.Session().Destroy(sessionID, nil)

	info, _, err := client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	require.NotNil(t, info)
	assert.Equal(t, api.SessionBehaviorDelete, info.Behavior,
		"Behavior must be 'delete'")
}

// CS-009: Test session with release behavior
func TestSessionReleaseBehavior(t *testing.T) {
	client := getClient(t)

	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name:     "release-behavior-" + randomID(),
		TTL:      "30s",
		Behavior: api.SessionBehaviorRelease,
	}, nil)
	require.NoError(t, err)
	defer client.Session().Destroy(sessionID, nil)

	info, _, err := client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	require.NotNil(t, info)
	assert.Equal(t, api.SessionBehaviorRelease, info.Behavior,
		"Behavior must be 'release'")
}

// CS-010: Test session KV lock — destroy with release behavior preserves key
func TestSessionDestroyReleasesLocks(t *testing.T) {
	client := getClient(t)

	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name:     "destroy-release-" + randomID(),
		TTL:      "30s",
		Behavior: api.SessionBehaviorRelease,
	}, nil)
	require.NoError(t, err)

	// Acquire KV lock
	key := "session-release-" + randomID()
	acquired, _, err := client.KV().Acquire(&api.KVPair{
		Key:     key,
		Value:   []byte("locked-value"),
		Session: sessionID,
	}, nil)
	require.NoError(t, err)
	require.True(t, acquired)
	defer client.KV().Delete(key, nil)

	// Verify locked
	pair, _, err := client.KV().Get(key, nil)
	require.NoError(t, err)
	require.NotNil(t, pair)
	assert.Equal(t, sessionID, pair.Session)

	// Destroy session — release behavior should release lock but preserve key
	_, err = client.Session().Destroy(sessionID, nil)
	require.NoError(t, err)

	time.Sleep(500 * time.Millisecond)

	// Key should still exist but lock should be released
	pair, _, err = client.KV().Get(key, nil)
	require.NoError(t, err)
	require.NotNil(t, pair,
		"Key must still exist after session destroy with release behavior")
	assert.Empty(t, pair.Session,
		"Lock must be released after session destroy")
	assert.Equal(t, []byte("locked-value"), pair.Value,
		"Value must be preserved after session destroy with release behavior")
}

// CS-011: Test session KV lock — destroy with delete behavior deletes key
func TestSessionDestroyDeletesKeys(t *testing.T) {
	client := getClient(t)

	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name:     "destroy-delete-" + randomID(),
		TTL:      "30s",
		Behavior: api.SessionBehaviorDelete,
	}, nil)
	require.NoError(t, err)

	// Acquire KV lock
	key := "session-delete-" + randomID()
	acquired, _, err := client.KV().Acquire(&api.KVPair{
		Key:     key,
		Value:   []byte("to-be-deleted"),
		Session: sessionID,
	}, nil)
	require.NoError(t, err)
	require.True(t, acquired)
	defer client.KV().Delete(key, nil)

	// Destroy session — delete behavior should delete the key
	_, err = client.Session().Destroy(sessionID, nil)
	require.NoError(t, err)

	time.Sleep(500 * time.Millisecond)

	// Key should be deleted
	pair, _, err := client.KV().Get(key, nil)
	require.NoError(t, err)
	assert.Nil(t, pair,
		"Key must be deleted after session destroy with delete behavior")
}

// CS-012: Test node sessions
func TestSessionNode(t *testing.T) {
	client := getClient(t)

	// Get node name
	self, err := client.Agent().Self()
	require.NoError(t, err)
	nodeName := self["Config"]["NodeName"].(string)
	require.NotEmpty(t, nodeName)

	// Create session
	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name: "node-session-" + randomID(),
		TTL:  "30s",
	}, nil)
	require.NoError(t, err)
	defer client.Session().Destroy(sessionID, nil)

	// List sessions for this node
	sessions, meta, err := client.Session().Node(nodeName, nil)
	require.NoError(t, err, "Session.Node should succeed")
	assert.Greater(t, meta.LastIndex, uint64(0))

	// Our session should be in the list
	found := false
	for _, s := range sessions {
		if s.ID == sessionID {
			found = true
			assert.Equal(t, nodeName, s.Node,
				"Session node must match queried node")
			break
		}
	}
	assert.True(t, found, "Created session must be found in node session list")
}

// CS-013: Test multiple sessions on same node
func TestSessionMultipleOnNode(t *testing.T) {
	client := getClient(t)

	var ids []string
	for i := 0; i < 5; i++ {
		id, _, err := client.Session().Create(&api.SessionEntry{
			Name: "multi-" + randomID(),
			TTL:  "30s",
		}, nil)
		require.NoError(t, err)
		ids = append(ids, id)
	}
	defer func() {
		for _, id := range ids {
			client.Session().Destroy(id, nil)
		}
	}()

	sessions, _, err := client.Session().List(nil)
	require.NoError(t, err)

	foundCount := 0
	for _, s := range sessions {
		for _, id := range ids {
			if s.ID == id {
				foundCount++
				break
			}
		}
	}
	assert.Equal(t, 5, foundCount,
		"All 5 created sessions must be found in list")
}

// CS-014: Test CreateNoChecks
func TestSessionCreateNoChecks(t *testing.T) {
	client := getClient(t)

	sessionID, _, err := client.Session().CreateNoChecks(&api.SessionEntry{
		Name: "no-checks-" + randomID(),
		TTL:  "30s",
	}, nil)
	require.NoError(t, err, "CreateNoChecks should succeed")
	require.NotEmpty(t, sessionID)
	defer client.Session().Destroy(sessionID, nil)

	info, _, err := client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	require.NotNil(t, info)
	assert.Equal(t, sessionID, info.ID)
}

// CS-015: Test periodic renewal
func TestSessionRenewPeriodic(t *testing.T) {
	client := getClient(t)

	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name: "periodic-" + randomID(),
		TTL:  "15s",
	}, nil)
	require.NoError(t, err)
	defer client.Session().Destroy(sessionID, nil)

	doneCh := make(chan struct{})
	errCh := make(chan error, 1)

	go func() {
		errCh <- client.Session().RenewPeriodic("15s", sessionID, nil, doneCh)
	}()

	// Let it renew once
	time.Sleep(2 * time.Second)

	// Session should still be alive
	info, _, err := client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	assert.NotNil(t, info,
		"Session must still be alive during periodic renewal")

	close(doneCh)
	err = <-errCh
	assert.NoError(t, err)

	// After close, RenewPeriodic destroys the session
	info, _, err = client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	assert.Nil(t, info,
		"Session must be destroyed after RenewPeriodic stops")
}

// ==================== Health-Check → Session Invalidation ====================

// CS-016: Test session auto-invalidation when linked check becomes Critical.
// Consul core behavior: sessions are automatically destroyed when their associated
// health checks fail. This is the safety mechanism for distributed locking —
// if a node holding a lock becomes unhealthy, its sessions are invalidated and
// locks are released so other nodes can acquire them.
func TestSessionInvalidatedOnCheckCritical(t *testing.T) {
	client := getClient(t)

	serviceID := "cs016-session-inval-" + randomID()
	checkID := "service:" + serviceID

	// Step 1: Register a service with a TTL check (starts passing)
	err := client.Agent().ServiceRegister(&api.AgentServiceRegistration{
		ID:   serviceID,
		Name: serviceID,
		Port: 8080,
		Check: &api.AgentServiceCheck{
			CheckID: checkID,
			TTL:     "30s",
			Status:  "passing",
		},
	})
	require.NoError(t, err, "Service registration should succeed")
	defer client.Agent().ServiceDeregister(serviceID)

	time.Sleep(500 * time.Millisecond)

	// Step 2: Create a session linked to this service check
	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name:          "check-linked-" + randomID(),
		TTL:           "60s",
		NodeChecks:    []string{},                // no node checks
		ServiceChecks: []api.ServiceCheck{{ID: checkID}},
	}, nil)
	require.NoError(t, err, "Session create should succeed")
	require.NotEmpty(t, sessionID)

	// Step 3: Acquire a KV lock with this session
	kvKey := "test/session-inval/" + randomID()
	acquired, _, err := client.KV().Acquire(&api.KVPair{
		Key:     kvKey,
		Value:   []byte("locked-by-session"),
		Session: sessionID,
	}, nil)
	require.NoError(t, err)
	require.True(t, acquired, "KV lock should be acquired")
	defer client.KV().Delete(kvKey, nil)

	// Verify session exists and lock is held
	info, _, err := client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	require.NotNil(t, info, "Session should exist before check failure")

	pair, _, err := client.KV().Get(kvKey, nil)
	require.NoError(t, err)
	require.NotNil(t, pair)
	assert.Equal(t, sessionID, pair.Session, "KV should be locked by session")

	// Step 4: Fail the check (mark as Critical)
	err = client.Agent().FailTTL(checkID, "node crashed")
	require.NoError(t, err, "FailTTL should succeed")

	// Wait for session invalidation to propagate
	time.Sleep(2 * time.Second)

	// Step 5: Verify session was automatically invalidated
	info, _, err = client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	assert.Nil(t, info,
		"Session must be automatically destroyed when linked check becomes Critical")

	// Step 6: Verify KV lock was released (release behavior is default)
	pair, _, err = client.KV().Get(kvKey, nil)
	require.NoError(t, err)
	if pair != nil {
		assert.Empty(t, pair.Session,
			"KV lock must be released after session invalidation")
	}
}

// CS-017: Test session with serfHealth — session should survive while node is healthy.
// This is the default session behavior: NodeChecks=["serfHealth"].
func TestSessionSurvivesWithHealthySerfCheck(t *testing.T) {
	client := getClient(t)

	// Create session with default checks (serfHealth)
	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name: "serf-healthy-" + randomID(),
		TTL:  "30s",
	}, nil)
	require.NoError(t, err)
	defer client.Session().Destroy(sessionID, nil)

	// Session should exist (serfHealth is passing)
	info, _, err := client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	require.NotNil(t, info, "Session should exist while serfHealth is passing")
	assert.Equal(t, sessionID, info.ID)

	// serfHealth should be passing
	checks, _, err := client.Health().State("passing", nil)
	require.NoError(t, err)
	found := false
	for _, c := range checks {
		if c.CheckID == "serfHealth" {
			found = true
			assert.Equal(t, api.HealthPassing, c.Status)
			break
		}
	}
	assert.True(t, found, "serfHealth check should exist and be passing")
}

// CS-018: Test session invalidation with delete behavior — KV keys should be deleted.
func TestSessionInvalidatedOnCheckCriticalDeleteBehavior(t *testing.T) {
	client := getClient(t)

	serviceID := "cs018-del-inval-" + randomID()
	checkID := "service:" + serviceID

	// Register service with TTL check
	err := client.Agent().ServiceRegister(&api.AgentServiceRegistration{
		ID:   serviceID,
		Name: serviceID,
		Port: 8081,
		Check: &api.AgentServiceCheck{
			CheckID: checkID,
			TTL:     "30s",
			Status:  "passing",
		},
	})
	require.NoError(t, err)
	defer client.Agent().ServiceDeregister(serviceID)

	time.Sleep(500 * time.Millisecond)

	// Create session with DELETE behavior linked to service check
	sessionID, _, err := client.Session().Create(&api.SessionEntry{
		Name:          "delete-inval-" + randomID(),
		TTL:           "60s",
		Behavior:      api.SessionBehaviorDelete,
		NodeChecks:    []string{},
		ServiceChecks: []api.ServiceCheck{{ID: checkID}},
	}, nil)
	require.NoError(t, err)
	require.NotEmpty(t, sessionID)

	// Acquire KV lock
	kvKey := "test/session-del-inval/" + randomID()
	acquired, _, err := client.KV().Acquire(&api.KVPair{
		Key:     kvKey,
		Value:   []byte("will-be-deleted"),
		Session: sessionID,
	}, nil)
	require.NoError(t, err)
	require.True(t, acquired)
	defer client.KV().Delete(kvKey, nil)

	// Fail the check
	err = client.Agent().FailTTL(checkID, "check failed")
	require.NoError(t, err)

	time.Sleep(2 * time.Second)

	// Session should be invalidated
	info, _, err := client.Session().Info(sessionID, nil)
	require.NoError(t, err)
	assert.Nil(t, info,
		"Session with delete behavior must be invalidated on check Critical")

	// KV key should be DELETED (not just released)
	pair, _, err := client.KV().Get(kvKey, nil)
	require.NoError(t, err)
	assert.Nil(t, pair,
		"KV key must be deleted when session with delete behavior is invalidated")
}

File: sdk-tests/batata-consul-client-tests/tests/session_test.rs
mod common;

use batata_consul_client::{
    AgentServiceCheck, AgentServiceRegistration, KVPair, ServiceCheck, SessionEntry,
};

#[tokio::test]
async fn test_session_create_and_destroy() {
    let client = common::create_client();

    let entry = SessionEntry {
        name: Some(format!("session-create-{}", common::test_id())),
        ttl: Some("30s".to_string()),
        ..Default::default()
    };

    // Create
    let (session_id, _) = client.session_create(&entry, &common::w()).await.unwrap();
    assert!(!session_id.is_empty(), "session ID should not be empty");

    // Verify it exists via info
    let (entries, _) = client
        .session_info(&session_id, &common::q())
        .await
        .unwrap();
    assert_eq!(
        entries.len(),
        1,
        "session_info should return exactly 1 entry"
    );
    assert_eq!(entries[0].id.as_deref(), Some(session_id.as_str()));

    // Destroy
    let _ = client
        .session_destroy(&session_id, &common::w())
        .await
        .unwrap();

    // Verify gone
    let (entries, _) = client
        .session_info(&session_id, &common::q())
        .await
        .unwrap();
    assert!(entries.is_empty(), "session should be gone after destroy");
}

#[tokio::test]
async fn test_session_info() {
    let client = common::create_client();
    let name = format!("session-info-{}", common::test_id());

    let entry = SessionEntry {
        name: Some(name.clone()),
        ttl: Some("30s".to_string()),
        ..Default::default()
    };
    let (session_id, _) = client.session_create(&entry, &common::w()).await.unwrap();

    let (entries, _meta) = client
        .session_info(&session_id, &common::q())
        .await
        .unwrap();
    assert_eq!(entries.len(), 1, "should return 1 session");

    let session = &entries[0];
    assert_eq!(session.id.as_deref(), Some(session_id.as_str()));
    assert_eq!(
        session.name.as_deref(),
        Some(name.as_str()),
        "session name should match"
    );
    assert!(session.node.is_some(), "session should have a node");
    assert!(
        !session.node.as_deref().unwrap_or("").is_empty(),
        "node should not be empty"
    );
    assert!(session.create_index > 0, "create_index should be positive");

    // Cleanup
    let _ = client.session_destroy(&session_id, &common::w()).await;
}

#[tokio::test]
async fn test_session_list() {
    let client = common::create_client();
    let mut session_ids = Vec::new();

    // Create 2 sessions
    for i in 0..2 {
        let entry = SessionEntry {
            name: Some(format!("session-list-{}-{}", i, common::test_id())),
            ttl: Some("30s".to_string()),
            ..Default::default()
        };
        let (id, _) = client.session_create(&entry, &common::w()).await.unwrap();
        assert!(!id.is_empty());
        session_ids.push(id);
    }

    // List all sessions
    let (entries, _) = client.session_list(&common::q()).await.unwrap();
    assert!(
        entries.len() >= 2,
        "should list at least 2 sessions, got {}",
        entries.len()
    );

    // Verify our sessions are in the list
    for sid in &session_ids {
        let found = entries
            .iter()
            .any(|e| e.id.as_deref() == Some(sid.as_str()));
        assert!(found, "session {} should be in the list", sid);
    }

    // Cleanup
    for sid in &session_ids {
        let _ = client.session_destroy(sid, &common::w()).await;
    }
}

#[tokio::test]
async fn test_session_node() {
    let client = common::create_client();

    let entry = SessionEntry {
        name: Some(format!("session-node-{}", common::test_id())),
        ttl: Some("30s".to_string()),
        ..Default::default()
    };
    let (session_id, _) = client.session_create(&entry, &common::w()).await.unwrap();

    // Get session info to find the node name
    let (entries, _) = client
        .session_info(&session_id, &common::q())
        .await
        .unwrap();
    assert_eq!(entries.len(), 1);
    let node_name = entries[0].node.clone().expect("session should have a node");
    assert!(!node_name.is_empty());

    // Query sessions by node
    let (node_sessions, _) = client.session_node(&node_name, &common::q()).await.unwrap();
    assert!(
        !node_sessions.is_empty(),
        "node should have at least one session"
    );

    let found = node_sessions
        .iter()
        .any(|e| e.id.as_deref() == Some(session_id.as_str()));
    assert!(
        found,
        "our session should appear in the node's session list"
    );

    // Cleanup
    let _ = client.session_destroy(&session_id, &common::w()).await;
}

#[tokio::test]
async fn test_session_renew() {
    let client = common::create_client();

    let entry = SessionEntry {
        name: Some(format!("session-renew-{}", common::test_id())),
        ttl: Some("30s".to_string()),
        ..Default::default()
    };
    let (session_id, _) = client.session_create(&entry, &common::w()).await.unwrap();

    // Renew the session
    let (renewed, _) = client
        .session_renew(&session_id, &common::w())
        .await
        .unwrap();
    assert!(!renewed.is_empty(), "renew should return the session entry");
    assert_eq!(
        renewed[0].id.as_deref(),
        Some(session_id.as_str()),
        "renewed session ID should match"
    );

    // Verify it is still alive
    let (entries, _) = client
        .session_info(&session_id, &common::q())
        .await
        .unwrap();
    assert_eq!(entries.len(), 1, "session should still exist after renew");

    // Cleanup
    let _ = client.session_destroy(&session_id, &common::w()).await;
}

#[tokio::test]
async fn test_session_create_no_checks() {
    let client = common::create_client();

    let entry = SessionEntry {
        name: Some(format!("session-no-checks-{}", common::test_id())),
        ttl: Some("30s".to_string()),
        ..Default::default()
    };

    let (session_id, _) = client
        .session_create_no_checks(&entry, &common::w())
        .await
        .unwrap();
    assert!(!session_id.is_empty(), "session ID should not be empty");

    // Verify the session exists and has the correct name
    let (entries, _) = client
        .session_info(&session_id, &common::q())
        .await
        .unwrap();
    assert_eq!(entries.len(), 1);
    assert_eq!(
        entries[0].name.as_deref(),
        Some(entry.name.as_deref().unwrap()),
        "session name should match"
    );

    // Cleanup
    let _ = client.session_destroy(&session_id, &common::w()).await;
}

#[tokio::test]
async fn test_session_with_behavior() {
    let client = common::create_client();

    let entry = SessionEntry {
        name: Some(format!("session-behavior-{}", common::test_id())),
        behavior: Some("delete".to_string()),
        ttl: Some("30s".to_string()),
        ..Default::default()
    };

    let (session_id, _) = client.session_create(&entry, &common::w()).await.unwrap();
    assert!(!session_id.is_empty(), "session ID should not be empty");

    // Verify the session exists with the specified behavior
    let (entries, _) = client
        .session_info(&session_id, &common::q())
        .await
        .unwrap();
    assert_eq!(entries.len(), 1);
    assert_eq!(
        entries[0].behavior.as_deref(),
        Some("delete"),
        "session behavior should be 'delete'"
    );

    // Cleanup
    let _ = client.session_destroy(&session_id, &common::w()).await;
}

// ==================== Health-Check → Session Invalidation ====================

/// Test session auto-invalidation when linked check becomes Critical.
///
/// Consul core behavior: sessions are automatically destroyed when their
/// associated health checks fail. This is the safety mechanism for distributed
/// locking — if a node holding a lock becomes unhealthy, its sessions are
/// invalidated and locks are released so other nodes can acquire them.
#[tokio::test]
async fn test_session_invalidated_on_check_critical() {
    common::init_tracing();
    let client = common::create_client();
    let suffix = common::test_id();
    let svc_id = format!("session-inval-svc-{}", suffix);
    let check_id = format!("service:{}", svc_id);

    // Step 1: Register a service with a TTL check (starts passing)
    let reg = AgentServiceRegistration {
        id: Some(svc_id.clone()),
        name: svc_id.clone(),
        port: Some(8080),
        check: Some(AgentServiceCheck {
            check_id: Some(check_id.clone()),
            ttl: Some("30s".to_string()),
            status: Some("passing".to_string()),
            ..Default::default()
        }),
        ..Default::default()
    };
    client
        .agent_service_register(&reg, &common::w())
        .await
        .unwrap();

    tokio::time::sleep(std::time::Duration::from_millis(500)).await;

    // Step 2: Create a session linked to this service check
    let entry = SessionEntry {
        name: Some(format!("check-linked-{}", suffix)),
        ttl: Some("60s".to_string()),
        node_checks: Some(vec![]),
        service_checks: Some(vec![ServiceCheck {
            id: check_id.clone(),
            namespace: None,
        }]),
        ..Default::default()
    };
    let (session_id, _) = client.session_create(&entry, &common::w()).await.unwrap();
    assert!(!session_id.is_empty());

    // Step 3: Acquire a KV lock with this session
    let kv_key = format!("test/session-inval/{}", suffix);
    let kv_pair = KVPair {
        key: kv_key.clone(),
        value: Some("locked-by-session".to_string()),
        session: Some(session_id.clone()),
        ..Default::default()
    };
    let (acquired, _) = client.kv_acquire(&kv_pair, &common::w()).await.unwrap();
    assert!(acquired, "KV lock should be acquired");

    // Verify session exists
    let (entries, _) = client
        .session_info(&session_id, &common::q())
        .await
        .unwrap();
    assert_eq!(entries.len(), 1, "Session should exist before check failure");

    // Step 4: Fail the check (mark as Critical)
    client
        .agent_check_fail(&check_id, "node crashed", &common::w())
        .await
        .unwrap();

    // Wait for session invalidation to propagate
    tokio::time::sleep(std::time::Duration::from_secs(2)).await;

    // Step 5: Verify session was automatically invalidated
    let (entries, _) = client
        .session_info(&session_id, &common::q())
        .await
        .unwrap();
    assert!(
        entries.is_empty(),
        "Session must be automatically destroyed when linked check becomes Critical"
    );

    // Step 6: Verify KV lock was released
    let (pair, _) = client.kv_get(&kv_key, &common::q()).await.unwrap();
    if let Some(pair) = pair {
        assert!(
            pair.session.as_deref().unwrap_or("").is_empty(),
            "KV lock must be released after session invalidation"
        );
    }

    // Cleanup
    let _ = client.kv_delete(&kv_key, &common::w()).await;
    let _ = client.agent_service_deregister(&svc_id, &common::w()).await;
}

/// Test session survives when serfHealth is passing (default behavior).
#[tokio::test]
async fn test_session_survives_with_healthy_serf_check() {
    let client = common::create_client();
    let suffix = common::test_id();

    // Create session with default checks (serfHealth)
    let entry = SessionEntry {
        name: Some(format!("serf-healthy-{}", suffix)),
        ttl: Some("30s".to_string()),
        ..Default::default()
    };
    let (session_id, _) = client.session_create(&entry, &common::w()).await.unwrap();

    // Session should exist (serfHealth is passing)
    let (entries, _) = client
        .session_info(&session_id, &common::q())
        .await
        .unwrap();
    assert_eq!(
        entries.len(),
        1,
        "Session should exist while serfHealth is passing"
    );
    assert_eq!(entries[0].id.as_deref(), Some(session_id.as_str()));

    // serfHealth should be passing
    let (checks, _) = client.health_state("passing", &common::q()).await.unwrap();
    let serf = checks.iter().find(|c| c.check_id == "serfHealth");
    assert!(serf.is_some(), "serfHealth check should exist and be passing");

    // Cleanup
    let _ = client.session_destroy(&session_id, &common::w()).await;
}
